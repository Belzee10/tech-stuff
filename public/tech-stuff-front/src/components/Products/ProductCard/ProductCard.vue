<template>
  <v-card class=" product-card">
    <v-img
      class="white--text align-end"
      height="150px"
      :src="image"
      lazy-src="https://picsum.photos/id/11/10/6"
    >
      <v-card-title>{{ name }}</v-card-title>
    </v-img>

    <v-card-subtitle class="pb-0">{{ brand }}</v-card-subtitle>

    <v-card-text class="text--primary">
      <span class="font-weight-bold">${{ price }}</span>
    </v-card-text>

    <v-card-actions>
      <div v-if="inStock > 0">
        <v-btn
          small
          block
          depressed
          class="btn-small"
          color="secondary"
          @click="$emit('remove-from-cart', id)"
        >
          -
        </v-btn>
        <v-btn
          small
          block
          depressed
          class="btn-small"
          color="secondary"
          @click="$emit('add-to-cart', id)"
        >
          +
        </v-btn>
      </div>
      <v-btn
        v-else
        small
        block
        depressed
        class="btn-full text-capitalize"
        color="secondary"
        @click="$emit('add-to-cart', id)"
      >
        <v-icon small class="mr-1">mdi-cart-plus</v-icon>
        Add to Cart
      </v-btn>
    </v-card-actions>
  </v-card>
</template>

<script>
export default {
  name: 'ProductCard',
  props: {
    id: {
      type: Number,
      default: null
    },
    name: {
      type: String,
      default: ''
    },
    price: {
      type: Number,
      default: null
    },
    image: {
      type: String,
      default: ''
    },
    brand: {
      type: String,
      default: ''
    },
    inStock: {
      type: Number,
      default: 0
    }
  }
};
</script>

<style lang="scss" scoped></style>
