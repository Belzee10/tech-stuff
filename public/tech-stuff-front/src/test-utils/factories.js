import { mount, createLocalVue } from '@vue/test-utils';
import mergeWith from 'lodash.mergewith';
import Vuetify from 'vuetify';
import Vuex from 'vuex';

const localVue = createLocalVue();
localVue.use(Vuex);

let vuetify;

beforeEach(() => {
  vuetify = new Vuetify();
});

const customizer = (_, srcValue) => {
  if (Array.isArray(srcValue)) {
    return srcValue;
  }
  if (srcValue instanceof Object && Object.keys(srcValue).length === 0) {
    return srcValue;
  }
};

/**
 * create component wrapper
 * @param {*} Component
 * @param {Object} overrides
 */
const createWrapper = (Component, overrides) => {
  const defaultMountingOptions = {
    localVue,
    vuetify
  };
  return mount(
    Component,
    mergeWith(defaultMountingOptions, overrides, customizer)
  );
};

/**
 * create Vuex store
 * @param {Object} overrides
 */
const createStore = overrides => {
  const defaultStoreConfig = {};
  return new Vuex.Store(mergeWith(defaultStoreConfig, overrides, customizer));
};

export { createWrapper, createStore };
