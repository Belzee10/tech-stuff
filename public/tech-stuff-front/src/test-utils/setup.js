import Vue from 'vue';
import Vuetify from 'vuetify';

Vue.config.productionTip = false;

Vue.use(Vuetify);

/**
 * silent "console.error"
 */
global.console = {
  error: jest.fn(),
  log: console.log,
  warn: console.warn,
  info: console.info,
  debug: console.debug
};
