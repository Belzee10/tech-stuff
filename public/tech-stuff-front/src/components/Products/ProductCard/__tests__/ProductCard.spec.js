import { createWrapper } from '@/test-utils/factories';
import ProductCard from '../ProductCard.vue';

describe('ProductCard.vue', () => {
  test('should emit a "add-to-cart" event', () => {
    const props = {
      id: '1'
    };
    const wrapper = createWrapper(ProductCard, {
      propsData: props
    });
    wrapper.find('.v-btn').trigger('click');
    expect(wrapper.emitted('add-to-cart')[0]).toEqual([props.id]);
  });
  test('should render two "btn-small", emit "add-to-cart" and "remove-from-cart" events', () => {
    const props = {
      id: '1',
      inStock: 2
    };
    const wrapper = createWrapper(ProductCard, {
      propsData: props
    });
    expect(wrapper.find('.btn-full').exists()).toBeFalsy();
    expect(wrapper.findAll('.btn-small')).toHaveLength(2);
    wrapper.find('.btn-small:first-child').trigger('click');
    expect(wrapper.emitted('remove-from-cart')[0]).toEqual([props.id]);
    wrapper.find('.btn-small:last-child').trigger('click');
    expect(wrapper.emitted('add-to-cart')[0]).toEqual([props.id]);
  });
});
