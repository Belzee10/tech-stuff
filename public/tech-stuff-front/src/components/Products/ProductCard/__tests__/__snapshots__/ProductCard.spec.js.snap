// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard.vue should render correctly 1`] = `
<div
  class=" product-card v-card v-sheet theme--light"
>
  <div
    class="v-responsive v-image white--text align-end"
    style="height: 150px;"
  >
    <div
      class="v-image__image v-image__image--preload v-image__image--cover"
      mode="in-out"
      name="fade-transition"
      style="background-image: url(https://picsum.photos/id/11/10/6); background-position: center center;"
    />
    <div
      class="v-responsive__content"
    >
      <div
        class="v-card__title"
      >
        
      </div>
    </div>
  </div>
   
  <div
    class="v-card__subtitle pb-0"
  >
    
  </div>
   
  <div
    class="v-card__text text--primary"
  >
    <span
      class="font-weight-bold"
    >
      $
    </span>
  </div>
   
  <div
    class="v-card__actions"
  >
    <button
      class="btn-full text-capitalize v-btn v-btn--block v-btn--depressed theme--light v-size--small secondary"
      type="button"
    >
      <span
        class="v-btn__content"
      >
        <i
          aria-hidden="true"
          class="v-icon notranslate mr-1 mdi mdi-cart-plus theme--light"
          style="font-size: 16px;"
        />
        
      Add to Cart
    
      </span>
    </button>
  </div>
</div>
`;
